### DISCUSSION TITLE
Release 2.0 & Roadmap - Proposals and Discussions

### DISCUSSION THREAD

**Discussion by sflynn850** [raceintospace/raceintospace#856] [General]
Hey all!

In my quest to be a useful member of the community, I wanted to centralize some discussions of Release 2.0, talk about what should be in it, and when we want to do that. 

# Wot I've Done / Roadmaps
* I set the Release 2.0 milestone date, completely arbitrarily, for October 1. I did this because the previously assigned date was in the past, and I figured it'd be easier to hit a date in the future than in the past.
* Created a Release 2.0.1 milestone with a date (again, arbitrarily) of November 15. I figure we can toss a few bugs in here.
* Created a Release 2.1 milestone with a date (arbitrarily! subject to change or veto or general complaining!) of January 15, 2025. I figure we'll toss some small enhancements in here.

My proposal is that we do (roughly) quarterly low-stress releases with an enhancement or two or five and/or some bugs. Not a huge deal to get something into a release or out of a release (we're all volunteers! this is fun!), but **regular releases will help folks see that this is a living project :)**

# TODO For 2.0

So this is where I'm a little in the woods — I'm not exactly sure what we need to do to releases, or where exactly we release _to_. It'd be neat to have that documented somewhere, so I'm either going to find that info in the tickets and consolidate it into a RELEASE.md, or create a new "help document release process" ticket.

## Add tickets to milestones

I'm gonna go through and see if I can associate merged code and its related tickets with our 2.0 Release milestone — so we can get a changelog out.

## Make a changelog

Suppose I ought to do that too, eh?

## Target Platforms

As I understand it, we're targeting Linux (Ubuntu, specifically), Windows 11, and MacOS 14+.

### MacOS
If we can get a MacOS build together, homebrew seems like the logical choice. There are a number of issues currently blocking a MacOS release: #781, #366, #347 and (maybe) #249. These probably all boil down to: The game doesn't build on MacOS right now. It _should_, though, and I'm gonna see what I can do to rectify that.

### Linux
There seems to be some desire to get this out in the Ubuntu package manager (#249), and I assume also Debian. I'm not aware of any Linux-specific issues blocking release.

### Windows binaries
I don't think we have any Windows-specific issues blocking a release, though there's been some exploration of an installer in #498. Seems like we need a place folks can download the binaries, though. Where should that be? Just here in GH?

## Related Issues

I did a bit of searching for release-related tickets, and I'm putting them here for centralization reasons:
* #416 - Preparing Release, from 2020
* #808 - Some brief discussion of getting a release together?
* #249 - Windows & MacOS installer discussion
* #498 - Windows compilation and installation progress

**Comment by sflynn850** [comment]
Update: I can't read; the quite extensive changelog is here: https://github.com/raceintospace/raceintospace/blob/master/ChangeLog.md

**Reply by peyre** [reply]
Yes, I've been maintaining the changelog, keeping it up to date periodically.

**Reply by peyre** [reply]
It's up to date now.

**Reply by sflynn850** [reply]
Also you can have it generate a changelog when you create a new release based on the included milestones :)

**Comment by peyre** [comment]
Well this is certainly ambitious!  I have no trouble with us going to a regular release schedule, provided people are on board with doing a release on a regular basis.  We've had some trouble putting out releases _at all_ here on GitHub, until recently.

Alsed is currently working on fixing the old music corruption issue (#209), and it would be ideal if we could include that in the 2.0 release - though I would hate to stall it out on account of one bug fix, so if it needs to go into 2.0.1, so be it.

I've always assumed we would release here on GitHub, but I'm a greenhorn at all this.  I can put up links etc. on the old SourceForge page, and maybe someone can think of another place to give it some publicity & easy availability.

We should also release on Flatpak, since a lot of the work for that is done and it makes the game more widely available.  For the Ubuntu package manager, ideally we could get it into the Snap repository.  That would be the replacement now for since the latest LTS release has switched to Snaps.  Having other packages available (RPM etc.) would be nice, but not important.

For historical reference, the earliest releases of the game (like .1, .2, etc.) were RPMs (or TARs) - but that was starting in 2007.  We switched over to DEBs with release .4.8 in 2010.  In fact Linux is the only OS that hasn't skipped a release, yet I'd say it's (ideally) our lowest priority since it's the smallest user base and the compile instructions actually work, at least for Ubuntu.  Also our Linux guy is willing, but pretty busy just now so we may not be able to count on him.

**Edit:** I've also looked into the possibility of maybe adding RIS to Steam - that'd get us out there.  However, they charge you $100 to put your game on Steam, even if you're giving it away!  Nuts to that.

**Reply by sflynn850** [reply]
Love it! I'm working on making sure everything that's gotten done for 2.0 is properly tagged and sorted (holy cow, SO MUCH has gotten done, it's truly amazing).

**Reply by peyre** [reply]
And you're doing quite a lot in a short time now.  Keep up the momentum!

**Reply by sflynn850** [reply]
It's amazing what you can do when your day job runs out of work for you!

**Reply by sflynn850** [reply]
@peyre Looks like we've got a fix in for #209 — so if you can verify that, I'll go ahead and make a 2.0beta2 release + tag, and we can have that ready to target when our release experts have a moment. Then, everything else we'll fix can go into 2.0.1 or 2.1.

Thoughts?

**Reply by peyre** [reply]
#209 seems to be fixed as far as I can tell.  There is a small glitch with #359 - the game shows at 4X (double size), but can't be turned down to 2X (old normal size) in the config file.  I think that one should probably be resolved first, though if others disagree I could waive that objection.

**Reply by sflynn850** [reply]
I concur with that – it'd be nice to have that working.

**Comment by sflynn850** [comment]
Ok! Everything referenced in the Changelog is now associated with the Release 2.0 milestone.

We appear to be squashing 75 bugs, adding 60 enhancements, and addressing a total of 151 issues. Neat stuff.

**Reply by peyre** [reply]
Well yes.  That represents a decade or so of work by different people off and on.  It also justifies going from 1.1 to 2.0 instead of 1.2.

**Comment by hweimer** [comment]
I think the biggest showstopper for a release is the ability to build distributable binaries. As far as I can see, the only thing that is working right now is the creation of binary tarballs for Linux, but nothing for Windows, MacOS, Linux distributions or services like Flatpak.

**Reply by sflynn850** [reply]
Agreed - @peyre says we've got a couple experts on that, so hoping they'll hop in soon.

**Comment by kkosciusz** [comment]
I have opened a PR here: https://github.com/raceintospace/raceintospace/pull/866
This should fix windows and macos builds, and provide a working way to generate windows installer.
Feel free to review that and provide comments.

**Reply by peyre** [reply]
It's merged.  Hope it helps!

**Reply by peyre** [reply]
Uh oh!  I just downloaded the source and tried to compile, and got this:

`CMake Error at CMakeLists.txt:2 (cmake_minimum_required):
  CMake 3.25 or higher is required.  You are running version 3.22.1
`

Must be that my version of Ubuntu (22.04 LTS) doesn't include 3.25+ in its repository.  Canonical has back the upgrade to 24.04 _ridiculously_ far back, more than halfway to the next release.  Maybe I should force an upgrade.

**Reply by kkosciusz** [reply]
I was hoping people will review this first, but hey, that's a fastest merge ever :)
Due to addition of vcpkg you will need to initialize the git submodule in your existing repo after `git pull`:
```shell
git submodule update --init --recursive
```

Building instructions were also updated in the README.md

**Reply by peyre** [reply]
Thanks for updating the readme.  I'm a little concerned that they might be a little vague for many people to follow (they always were), but it's good to have them updated.

I didn't check the time you made the PR; I just figured it would be good to merge your changes.  Safe too, no?  We can revert them if we need to, I assume.

**Reply by kkosciusz** [reply]
Yes, you can always revert. Feel free to proofread, test and expand the instructions!

**Comment by sflynn850** [comment]
Happy to review & test, esp for macos, but I probably won't get to it until
tonight or possibly Monday. (It's father's day here, and we have a full
weekend!)

**Reply by peyre** [reply]
Oh yeah it is, isn't it?  Ha, I'll be taking my kid to the dentist in a couple hours.  Meanwhile I'm upgrading my OS so I can hopefully compile the game again to test it myself.

**Comment by peyre** [comment]
I got my OS upgraded yesterday and was able to compile the game, but when I tried this morning with Krzystof's latest change, I got the following.  Not sure what I need to do to fix this.

> CMake Error at src/game/CMakeLists.txt:96 (find_package):
>   By not providing "Findcereal.cmake" in CMAKE_MODULE_PATH this project has
>   asked CMake to find a package configuration file provided by "cereal", but
>   CMake did not find one.
> 
>   Could not find a package configuration file provided by "cereal" with any
>   of the following names:
> 
>     cerealConfig.cmake
>     cereal-config.cmake
> 
>   Add the installation prefix of "cereal" to CMAKE_PREFIX_PATH or set
>   "cereal_DIR" to a directory containing one of the above files.  If "cereal"
>   provides a separate development package or SDK, be sure it has been
>   installed.
>

**Reply by kkosciusz** [reply]
Can you provide the following information?

- What's your system
- Did you clone into a new repo? If so, did you use `--recurse-submodules`?
- Or did you update your existing local repo? If so, did you use `git submodule update --init --recursive`?
- What cmake preset did you specify?

**Comment by peyre** [comment]
Oh sorry!  I'm running Xubuntu 24.04.  I cloned the repo with this command:

> cd ~/ && sudo rm -r raceintospace && git clone https://github.com/raceintospace/raceintospace

Should I change that to && git clone --recurse-submodules ?

**Reply by kkosciusz** [reply]
Yes. The repo now includes a submodule with vcpkg that is necessary for the build system to work. Pease let me know if anything in README.md is unclear, I'll make necessary corrections if needed.

**Reply by peyre** [reply]
:(  I'm still getting

> CMake Error at src/game/CMakeLists.txt:96 (find_package):
>   By not providing "Findcereal.cmake" in CMAKE_MODULE_PATH this project has
>   asked CMake to find a package configuration file provided by "cereal", but
>   CMake did not find one.
> 
>   Could not find a package configuration file provided by "cereal" with any
>   of the following names:
> 
>     cerealConfig.cmake
>     cereal-config.cmake
> 
>   Add the installation prefix of "cereal" to CMAKE_PREFIX_PATH or set
>   "cereal_DIR" to a directory containing one of the above files.  If "cereal"
>   provides a separate development package or SDK, be sure it has been
>   installed.

Do I need to add something to my compiling command?  This is what I do at present:

> cd ~/ && sudo rm -r raceintospace-build && mkdir raceintospace-build && cd raceintospace-build && cmake ../raceintospace && make && sudo make install && cd src/game && ./raceintospace

**Reply by kkosciusz** [reply]
You will need to check the README.md, the compilation instructions were changed.
Please use the `linux-release` preset:
```bash
cmake --preset linux-release
cmake --build --preset linux-release --target run
```

**Reply by peyre** [reply]
Ah!  You're right, I was missing some dependencies.  It's working now, though it's defaulting to 4xscale and the music is stuttery.  I'll create an Issue for the stuttering.

**Comment by alsed** [comment]
Hi everyone, i've succesfully build a flatpack from the master, so the flatpack manifest is ready for launch.
I need to pull it yet, and someone (aka @peyre) will have to also pull it to flathub ;)

I think i can create a tarball also, so we can release the latest version on github and link the flatpack build from there (in order not to build directly from master and also to include the music folder), but I don't know which version will be, since 2.0.0rc3 is already taken.

![Captura desde 2024-07-24 16-54-51](https://github.com/user-attachments/assets/e6957097-1ad1-47c2-ab18-cf6f04ba51f5)

**Reply by peyre** [reply]
Personally I think the version should show as just "2.0".

I'd be happy to push it out to Flatpak but I actually don't know how.

**Reply by hweimer** [reply]
Someones needs to file a PR against https://github.com/flathub/org.raceintospace.Raceintospace with the updated manifest. I think it's @pemensik who can merge it.

**Reply by alsed** [reply]
> Someones needs to file a PR against https://github.com/flathub/org.raceintospace.Raceintospace with the updated manifest. I think it's @pemensik who can merge it.

I can push the changes into the flatpak repository, but first we need a release with the 2.0 tag, which the manifest can reference to download.  The version can be only the source files in tar and zip formats (the contrib and test folder can be removed for cleanup, and include the music folder). I think it is also necessary to change the version in cmake (and in the metainfo file) from 2.0-rc3 to 2.0.0

**Comment by alsed** [comment]
Quick update: 
After more hours of testing than should be healthy, I've created a first snap version.
More testing is needed , but I think I'll upload the yaml manifest in the contrib folder, in case anyone wants to test it as well.

![Captura desde 2024-08-19 20-38-39](https://github.com/user-attachments/assets/a032195f-5228-4d97-88f3-40889942b1a1)

**Reply by peyre** [reply]
So this might be getting into the Ubuntu repository?  Yay!

**Reply by alsed** [reply]
Yes! But as with flatpak, in order for these to work, they need a link to the source files, so a release in github for 2.0.0 is pre-requisite.

**Comment by kkosciusz** [comment]
I have an access to a MacBook now, and I'm happy to report that I was able to build and launch the game on ARM hardware with a pristine clone of a repo on a first try!

Macos version seems to ignore OGG files and tries to play MIDI files, which have to be downloaded separately. But once they are copied to a correct directory, the music also works.

I'm not sure if we have ever distributed a Macos app bundle, but nowadays this would also require building intel and arm versions (universal app) and distributing MIDI files is not possible without license. Just saying this might need additional work if someone is willing to pick this up.

**Reply by alsed** [reply]
For some reason, the macOS version of RIS has it's own custom source files to handle the music. They are definitely outdated.
The sound files are in ogg format, I found it strange that can also play  the midi files for the music. That is not possible (at least to my knowledge) with the SDL-1 multimedia library.
I dont' have access to a mac, but if we consider a necessary step in order to release 2.0, I'm willing to help in any way I can.

**Reply by peyre** [reply]
I think we can proceed without the Mac release.  We can always release that once it's ready.

**Reply by dtseto** [reply]
I have a MacBook and Xcode so I'm going to have to ask you how you did that!

**Comment by hweimer** [comment]
With #908 being fixed, I don't see any remaining bugs that need to be addressed before doing a release. To proceed, I've created a file Release.md to document which steps are necessary for creating a release. I have working instructions for creating a source tarball and a generic Linux binary release, with more hopefully coming soon. I probably can get DEB/RPM to work (and maybe even Windows), but for the rest I will need some help.

**Reply by peyre** [reply]
That's great!  Pace Willisson did the Linux releases on the old Sourceforge code, and I asked if he'd be able to do them for us again.  He said sure, but he's been very busy and I haven't heard from him since, so it looks as though we can't count on him - so it's good you're covering the Linux side. Krzysztof seems to be occupied as well, but hopefully things will clear up for him soon and we can finally push out a release.  The game is **so** much better than 1.1, it seems a shame to not make it widely available.

**Reply by hweimer** [reply]
The idea behind the release instructions is that making a release becomes as easy as compiling the game, so you don't need specific persons for it.

If you agree, I will file a PR bumping the version to 2.0.0 and put up Linux and Windows binaries in my fork, so you can grab them from there and do the official release. I can also start drafting a press release.

**Reply by peyre** [reply]
That would be fantastic!  I've always wanted to be able to do the releases myself.  We'd have had many by now if I were able to do them.

**Reply by hweimer** [reply]
If you want to do the release yourself, that would probably be an even better idea. That way, we can test whether the release instructions I've put up are actually useful.

(To bump the version to 2.0.0, you need to remove the "rc3" from the CMakeLists.txt in the top directory.)

**Reply by peyre** [reply]
You're saying it's ready for me right now?

**Reply by hweimer** [reply]
I would say so. I've been able to build a source tarball, a generic Linux tarball, DEB/RPM packages, and a Windows installer binary. I think that is sufficient for a release, if we manage to get additional platforms working later on, we can still add them when they are ready.

(The Windows installer even works flawlessly under Wine. Not that this is important, but I wanted to make sure that the game runs outside the virtual machine it was built on.)

**Reply by peyre** [reply]
Wow!  I followed your instructions for creating Linux binaries and it seems to have successfully created a tarball and a .deb.  However, the RPM one failed quickly:

> CPack: Create package using RPM
> CPack: Install projects
> CPack: - Run preinstall target for: raceintospace
> CPack: - Install project: raceintospace []
> CPack: Create package
> CMake Error at /usr/share/cmake-3.28/Modules/Internal/CPack/CPackRPM.cmake:832 (message):
  RPM package requires rpmbuild executable
> Call Stack (most recent call first):
  /usr/share/cmake-3.28/Modules/Internal/CPack/CPackRPM.cmake:1982 (cpack_rpm_generate_package)
> 
> CPack Error: Error while execution CPackRPM.cmake
> CPack Error: Problem compressing the directory
> CPack Error: Error when generating package: raceintospace

Also, is it possible to add some description to the binary?  When I open the .deb, it doesn't say what this is, really.

![cmake](https://github.com/user-attachments/assets/3829c87c-8bb2-4180-bcb7-ddffa7e830f6)

**Reply by hweimer** [reply]
Good catch! You need to install the rpm package, I forgot to mention that.

I will see how we can add more information to the packages.

**Reply by peyre** [reply]
Ah! That did it.  **sudo apt install rpm**, and then it built an rpm of about the right size.  Brilliant!  I'm a little uncertain about how to approach the Windows side...

**Reply by peyre** [reply]
Hey, what you mentioned about CMakeLists.txt - is that line 7, **set (raceintospace_VERSION_EXTRA rc3)**?  Do I just take out the "rc3", changing it to set (raceintospace_VERSION_EXTRA), or what?

**Reply by hweimer** [reply]
Exactly. Then, the version string will become 2.0.0.

Building the Windows installer shouldn't be too difficult. I managed to do it, despite not having done any programming under Windows for more than 20 years. Just grab Visual Studio, git, and NSIS, clone the repo, run the two cmake commands given in README.md, and finally run cpack.

If you need any assistance, I'm happy to help.

**Reply by peyre** [reply]
It should be noted that my desktop already had the prerequisites to compile the game, so some of that may be necessary for building a .deb as well.

I tried the Windows install.  Even though I told cmake to add itself to the $PATH, running the cmake commands in git complains that it can't find it.  I changed it to "C:\Program Files\CMake\bin\cmake..." and that let the first command run successfully, but the second errored out:

> CMake Error: CMake was unable to find a build program corresponding to "Ninja".  CMAKE_MAKE_PROGRAM is not set.  You probably need to select a different build tool.
> CMake Error: CMAKE_C_COMPILER not set, after EnableLanguage
> CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage
> -- Configuring incomplete, errors occurred!

**Reply by hweimer** [reply]
Did you install CMake using the Visual Studio Installer? And did you use the Visual Studio Developer Command Prompt (i.e., not cmd.exe)? Doing so gave me a shell where cmake was already in the path. Ninja should be there, too.

**Reply by peyre** [reply]
Oh no, I downloaded the cmake .msi.  The VS installer didn't seem to install cmake (maybe I missed that option in the install program?).

**Reply by hweimer** [reply]
You can run the Visual Studio Installer after initial installation and select additional components. It's a separate application.

**Reply by peyre** [reply]
Ok, I removed cmake and Visual Studio Installer, than reinstalled the latter, but cmake still isn't in my $PATH - or anywhere else I can find, for that matter.

**Reply by hweimer** [reply]
Did you install CMake using the Visual Studio Installer? If you then open Visual Studio and go to Tools/Visual Studio Command Prompt, you should get a shell where both cmake and ninja are already in your path.

**Comment by kkosciusz** [comment]
FYI, the windows installer is built automatically in the github CI
workflow. You only need to grab the exe file.

wt., 4 mar 2025, 06:46 użytkownik Hendrik W. ***@***.***>
napisał:

**Reply by hweimer** [reply]
I can't seem to access the CI builds (I only see that they have run), but I think they don't contain the music files.

**Comment by kkosciusz** [comment]
I'm not sure why I left the windows installer upload commented out, but it
can be re-enabled by uncommenting the line here:
https://github.com/raceintospace/raceintospace/actions/runs/13658987543/workflow#L17

As for distributing prebuilt rpm and deb packages, I'm not sure if that
will be useful for the linux users who download the game.
Wouldn't they need a matching distro/dependencies with what the build
targets? Probably that would be best left for distro maintainers.
We could target flatpak or snap, but I'm not really up to date with these
channels and what is required to target those.

wt., 4 mar 2025 o 12:57 Hendrik W. ***@***.***> napisał(a):

**Reply by hweimer** [reply]
I checked the sonames of the libraries, they seemed to be consistent over recent versions of Debian/Ubuntu/Fedora, so the packages should at least cover the big ones. The deb packages also contain information on dependencies, so they are a bit more useful than the tarball.

Ultimately, of course, we would like to have proper packages inside the distros. But in reality, I doubt there's going to be such a thing as a distro maintainer, it's probably going to be one of us pouring things downstream. Like how Petr Mensik has been doing it for Fedora. But that's something to address after the release if you ask me.

**Comment by sflynn850** [comment]
Okay, I've got a working MacOS binary for Apple silicon. Is the best thing to do just drop it here, or in the release, or...?

**Reply by hweimer** [reply]
Fantastic! Is this an installer file created using cpack?

**Reply by sflynn850** [reply]
So, there doesn't seem to be a macos release build target; we only have `-release` for windows and linux. However, the binary _should_ execute without any installer.

**Reply by hweimer** [reply]
The -release preset only controls whether debug information is included in the build or not. So, it's not really important. But I guess having a proper package ensures that all game data files and required libraries are included.

**Comment by sflynn850** [comment]
I went ahead and created a draft v2.0 release, so that I could have somewhere to stick the MacOS binary. I'm going to the woods tomorrow and will be quite unavailable for awhile, so didn't want to hold up the process.

**Comment by hweimer** [comment]
I just realized that it is also necessary to tag the release in git before packaging the release, as otherwise the title of the window still calls it "2.0.0rc3".

**Reply by sflynn850** [reply]
Well that's aggravating. The release format wants us to create the tag when we publish the release, which makes a certain kind of sense.

**Reply by sflynn850** [reply]
<img width="1014" alt="image" src="https://github.com/user-attachments/assets/53060638-3bad-4cd9-a9a1-ec84bc630d8e" />

**Reply by hweimer** [reply]
You can tag a particular state, build the binaries, and use this tag for the release in GitHub.
